src/sortby.ts

```diff
@@ -0,0 +1,95 @@
+/**
+  * Sort array by a specific element property, its value type must be one of string, number and date
+  * @param {string | undefined} sortByProperty specify a property from each element that sorting will be based on, undefined means sorty by element itself
+  * @param {
+      isDescending: boolean;
+      sortByPropertyType: string;
+    } options 
+  * @returns script version
+  */
+export function SortBy<T>(sortByProperty: string | symbol | number, options: {
+  isDescending: boolean;
+  type: string;
+} = {
+  isDescending: true,
+  type: 'string'
+}) {
+  const cachedValueKey = Symbol();
+  return function(
+    target: any,
+    propertyKey: string,
+    descriptor?: PropertyDescriptor
+  ) {
+    Object.defineProperty(target, propertyKey, {
+      set: function(value: Array<T>) {
+        
+        if (!value || !Array.isArray(value)) {
+          throw `Value of property ${propertyKey} is not a valid array!`;
+        }
+
+        if (!!sortByProperty && value.some(item => !item.hasOwnProperty(sortByProperty))) {
+          throw `Property ${propertyKey} has some elements that do not have property ${sortByProperty ? sortByProperty.toString() : ''}!`;
+        }
+      
+  
+        const errorMsg = (item: any, type: string) => `Value ${item} is not of predefined type ${type}`;
+
+        // Check if any element that does not follow pre-defined type
+        value.forEach((item: any) => {
+          switch(options.type) {
+            case 'string':
+              if (sortByProperty ? typeof item[sortByProperty] !== 'string' : typeof item !== 'string') {
+                throw errorMsg(item, options.type);
+              }  
+              return;
+            case 'number':
+              if (sortByProperty ? typeof item[sortByProperty] !== 'number' : typeof item !== 'number') {
+                throw errorMsg(item, options.type);
+              }
+              return;
+            case 'date':
+              if (sortByProperty ? !Date.parse(item[sortByProperty]) : !Date.parse(item)) {
+                throw errorMsg(item, options.type);
+              }
+              return; 
+            default:
+              throw errorMsg(item, options.type);
+          }
+        })
+
+        const sortFunc = (a: any, b: any) => {
```

Comment:
Move this function to the file level, it's static, no point of recreating it every time we use the decorator.
Also, you can apply the sorting direction while calling this function, if it's descending just multiply the result by `-1` then we save the call for reverse 😄 
What about nullish values? (if `a` or `b` are `undefined` or `null`) I would expect them to be pushed to the end.

Reply:
Also makes a lot of sense, code is much cleaner after adopting above comment, really love it :)

In addition, nullish values are also considered and there is a test case which deliberately tests it.